### DISCUSSION TITLE
NestJS Schematics

### DISCUSSION THREAD

**Discussion by tayambamwanza** [cyrilletuzi/vscode-angular-schematics#402] [Ideas]
# Checks before posting an issue

- [x] I understand I am writing to a *human being*, doing this open source project on his *free* time
- [x] I will be involved in my own issue, by answering questions and informing if the issue is resolved

# Configuration

- VS Code version: 1.52.1
- Angular version (`ng version`): 11.0.0

# Description of the issue

Hi this is more of a feature request than a bug report, I would be willing to assist or even undertake bringing this feature to fruition.

What I am asking is if it would be possible to also have a subset of schematics for generating nestjs files.

In the same way that there is a subset for ionic and material components, there could be a @nestjs/schematics option.

There reason why I feel this feature would be great is because Nestjs is a great back-end framework to use for angular (since it is inspired by it) and is definitely growing in popularity. It may be possible to re-use existing functionality of the extension to hook into the nestjs cli with less effort than if it might have had to be done from scratch.

**Comment by cyrilletuzi** [comment]
Hi @tayambamwanza,

While it could be possible to implement NestJS support in this extension, it seems there is already another VS Code extension for NestJS: [https://marketplace.visualstudio.com/items?itemName=AbhijoyBasak.nestjs-files](https://marketplace.visualstudio.com/items?itemName=AbhijoyBasak.nestjs-files)

Doesn't this extension work? (I'm not a NestJS user myself)

**Comment by tayambamwanza** [comment]
Thanks for bringing it to my attention, I can use it, although it would be nice to have support for it too (instead of two separate extensions), for example under the "Generate another schematic" option.

If you feel it's out of scope for this project then that's fine, I'll close the issue.

I think Nestjs is a natural back-end for Angular, because the concepts are borrowed from angular (dependency injection, services, modules etc). So i felt it does fit in this project to have the option to generate Nest files too. I would say ~80% of Angular knowledge transfers over to it making it very easy to learn for an Angular dev.

This was the video I watched when I first learned of it: [Leveraging Angular in Node.js with NestJS by Kamil Mysliwiec](https://www.youtube.com/watch?v=G6Bll03gVUY)

**Comment by cyrilletuzi** [comment]
If the other extension is working well, I would say it may be better to stick to it.

I quickly checked, and unfortunately it's more than just adding the `@nestjs/schematics` collection to the defaults, it would also require too:
- detect and manage NestJS config file instead of `angular.json` (seems easy stated like that but it is what would require the most work)
- when formatting the final command to launch, it must use `nest g` instead of `ng g` (otherwise the schematics work but the files are not generated in the right folder)

While all of this is technically possible, it would add quite some mess to this extension, which I feel not worth it if there is already another working extension.

I won't close the issue but move it to discussions (once it's possible, the feature seems unavailable currently), so other people can give their opinion.

**Comment by tayambamwanza** [comment]
If it's that much work than I understand, I didn't know the internals too well, i'm ok to just use the other extension.

**Comment by satoru260** [comment]
I think this is still a valid feature request. The unfortunate part about the other extension is that the user experience is very limited, and for some reason it doesn't add the generated service/controller to the module in the directory I click; It always updates the `app.module`, I have to manually delete the imports from there and import them in the correct module. At that point I might as well use snippets and create the files myself. Please add `@nestjs/schematics` support. I think the flow would be more consistent for fullstack devs and having a single extension for both front and back end development would be much easier, muscle memory and all that.

Another slight inconvenience with the other extension is that it adds all the schematics in the context menu, so every right click on folders has the entire schematics collection options.

**Comment by tayambamwanza** [comment]
@cyrilletuzi  @DavidTheProgrammer 

I dealt with this by using the option to add additional packages

```json
"ngschematics.schematics": [
  "@nestjs/schematics"
]
```

It detects the options correctly but for some reason it doesn't pick up the directory structure, this may be another option worth putting as a default, or do you feel it's out of scope because of the different technologies, I haven't faced and serious issues, only as mentioned before, it adds generated components to the wrong app.module.

**Reply by cyrilletuzi** [reply]
The path detection feature in this extension is currently based on the Angular CLI behavior (projects paths defined in `angular.json`, some fixed folders like `app`/`lib`, etc.).

Given that NestJS has a complete different behavior and this extension currently does not support it, issues in path management are expected.

**Reply by tayambamwanza** [reply]
Understandable, either way just glad I can make the schematics available via settings. thanks that awesome feature!

**Comment by cyrilletuzi** [comment]
~~Hi @DavidTheProgrammer, for information, an extended version of the extension is now available: **Schematics Pro**. It has more advanced Angular features, but is also framework agnostic, which allowed me to add Nest schematics (and also other frameworks schematics like React, Vue, etc.). So now with Schematics Pro you can have a single tool for all your projects. You will find all the details on the [dedicated website](https://cyrilletuzi.gumroad.com/l/schematicspro/). Cheers.~~

NestJS schematics are removed for now. They may be reintroduced later.
